api/base/filters.py

```diff
@@ -20,53 +30,184 @@ def filter_queryset(self, request, queryset, view):
             return queryset.sort(*ordering)
         return queryset
 
-query_pattern = re.compile(r'filter\[\s*(?P<field>\S*)\s*\]\s*')
 
+class FilterMixin(object):
+    """ View mixin with helper functions for filtering. """
+
+    QUERY_PATTERN = re.compile(r'^filter\[(?P<field>\w+)\](\[(?P<op>\w+)\])?$')
 
-def query_params_to_fields(query_params):
-    return {
-        query_pattern.match(key).groupdict()['field']: value
-        for key, value in query_params.items()
-        if query_pattern.match(key)
+    MATCH_OPERATORS = ('contains', 'icontains')
+    MATCHABLE_FIELDS = (ser.CharField, ser.ListField)
+
+    DEFAULT_OPERATOR = 'eq'
+    DEFAULT_OPERATOR_OVERRIDES = {
+        ser.CharField: 'icontains',
+        ser.ListField: 'contains',
     }
 
+    NUMERIC_FIELDS = (ser.IntegerField, ser.DecimalField, ser.FloatField)
 
-class FilterMixin(object):
-    """ View mixin with helper functions for filtering. """
+    DATE_FIELDS = (ser.DateTimeField, ser.DateField)
+    DATETIME_PATTERN = re.compile(r'^\d{4}\-\d{2}\-\d{2}(?P<time>T\d{2}:\d{2}(:\d{2}(:\d{1,6})?)?)$')
 
-    DEFAULT_OPERATOR = 'eq'
+    COMPARISON_OPERATORS = ('gt', 'gte', 'lt', 'lte')
+    COMPARABLE_FIELDS = NUMERIC_FIELDS + DATE_FIELDS
+
+    LIST_FIELDS = (ser.ListField, )
 
     def __init__(self, *args, **kwargs):
         super(FilterMixin, self).__init__(*args, **kwargs)
         if not self.serializer_class:
             raise NotImplementedError()
 
-    def is_filterable_field(self, key):
-        try:
-            return key.strip() in self.serializer_class.filterable_fields
-        except AttributeError:
-            return key.strip() in self.serializer_class._declared_fields
+    @staticmethod
+    def _isinstance_any(instance, bases):
+        return any([isinstance(instance, base) for base in bases])
+
+    def _get_default_operator(self, field):
+        return self.DEFAULT_OPERATOR_OVERRIDES.get(type(field), self.DEFAULT_OPERATOR)
+
+    def _get_valid_operators(self, field):
+        if self._isinstance_any(field, self.COMPARABLE_FIELDS):
+            return self.COMPARISON_OPERATORS + (self.DEFAULT_OPERATOR, )
+        elif self._isinstance_any(field, self.MATCHABLE_FIELDS):
+            return self.MATCH_OPERATORS + (self.DEFAULT_OPERATOR, )
+        else:
+            return None
+
+    def _get_field_or_error(self, field_name):
+        """
+        Check that the attempted filter field is valid
+
+        :raises InvalidFilterError: If the filter field is not valid
+        """
+        if field_name not in self.serializer_class._declared_fields:
+            raise InvalidFilterError(detail="'{0}' is not a valid field for this endpoint.".format(field_name))
+        if field_name not in getattr(self.serializer_class, 'filterable_fields', set()):
+            raise InvalidFilterFieldError(attribute=field_name)
```

Comment:
`InvalidFilterFieldError` does not accept a kwarg called attribute. Did you mean "value"?